Todo App is Finished

alijah-etch opened the pull request:
App added to repository. Tour of heroes tutorial done and ready for review before being committed to main.

alijah-etch committed:
Repository created

App added to repository

gavmck commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,4 @@
+{
+  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=827846
```
I guess these are autogenerated, but `//` comments aren't valid JSON. Not sure any comments are valid JSON.

gavmck commented on .vscode/launch.json:
```diff
@@ -0,0 +1,20 @@
+{
+  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Same comments comment

gavmck commented on .vscode/launch.json:
```diff
@@ -0,0 +1,20 @@
+  "version": "0.2.0",
+  "configurations": [
+    {
+      "name": "ng serve",
```
This is an interesting setup, does it boot the project up when you open it in vscode?

alijah-etch replied:
Yeah when I run the ng serve --open command it runs through vscode.

gavmck commented on src/app/app.component.css:
```diff
@@ -0,0 +1,21 @@
+/* AppComponent's private CSS styles */
```
Don't think you need to explain what these files are for as it should be inferred by the folder and filename

gavmck commented on src/app/app.component.css:
```diff
@@ -0,0 +1,21 @@
+/* AppComponent's private CSS styles */
+h1 {
+    margin-bottom: 0;
+  }
```
Tabbing's gone a bit wonky in here

gavmck commented on src/app/dashboard/dashboard.component.css:
```diff
@@ -0,0 +1,54 @@
+  max-width: 1000px;
+
+  /* flexbox */
+  display: -webkit-box;
```
Shouldn't need to prefix these, the autoprefixer will sort it out. No need to comment as flexbox either.

gavmck commented on src/app/dashboard/dashboard.component.css:
```diff
@@ -0,0 +1,54 @@
+}
+
+a {
+  background-color: #3f525c;
```
It would be good to pull these hardcoded values for colour, font sizing, spacing out into a set of design token variables.

JoshTheWanderer replied:
@gavmck it'd probably help to explain what design token variables are here.

gavmck commented on src/app/dashboard/dashboard.component.html:
```diff
@@ -0,0 +1,9 @@
+<h2>Top Heroes</h2>
+<div class="heroes-menu">
+    <a *ngFor="let hero of heroes"
```
Tabbing's a bit wonky in here too

gavmck commented on src/app/hero-detail/hero-detail.component.css:
```diff
@@ -0,0 +1,25 @@
+    /* HeroDetailComponent's private CSS styles */
```
Tabbing and comment again

gavmck commented on src/app/dashboard/dashboard.component.css:
```diff
@@ -0,0 +1,54 @@
+  text-align: center;
+}
+
+.heroes-menu {
```
This  should either be classed as part of the dashboard component (We use [BEM](https://getbem.com/), so `.dashboard__heroes-menu`) or moved out to its own component. Keeping the Block class matching the filename helps to locate code from the browser.

gavmck commented on src/app/hero-detail/hero-detail.component.css:
```diff
@@ -0,0 +1,25 @@
+        padding: .5rem;
+    }
+    button {
+        margin-top: 20px;
```
Why the mix of px vs rem?

gavmck commented on src/app/hero-detail/hero-detail.component.html:
```diff
@@ -0,0 +1,10 @@
+    <div><span>id: </span>{{hero.id}}</div>
+    <div>
+        <label for="hero-name">Hero name: </label>
+        <input id="hero-name" [(ngModel)]="hero.name" placeholder="Hero name"/>
```
needs an input type

gavmck commented on src/app/hero-detail/hero-detail.component.html:
```diff
@@ -0,0 +1,10 @@
+        <label for="hero-name">Hero name: </label>
+        <input id="hero-name" [(ngModel)]="hero.name" placeholder="Hero name"/>
+    </div>
+    <button type="button" (click)="save()">save</button>
```
It'd be nice if this submitted a form rather than using the button click event so we could take advantage of the keyboard enter event to submit as well.

gavmck commented on src/app/hero-search/hero-search.component.html:
```diff
@@ -0,0 +1,12 @@
+<div id="search-component">
+    <label for="search-box">Hero Search</label>
+    <input #searchBox id="search-box" (input)="search(searchBox.value)" />
```
needs an input type

gavmck commented on src/app/hero-search/hero-search.component.css:
```diff
@@ -0,0 +1,47 @@
+    /* HeroSearch private styles */
```
Think we could clean up the magic numbers, BEM this, maybe split it up

DanWebb replied:
When Gav says "magic numbers" he means numbers like `1.2rem` or `.5rem`

Create too many instances of these and you'll end up with a lack of consistent sizes/spacing across your app, or be using more effort than necessary to update all the various files in an attempt to keep it consistent, one approach we use to avoid this is using a centralised set of  [variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties). These are sometimes referred to as "design tokens".

For example `1.2rem` seems to be consistently used so it could be pulled out into a `var(--font-size-lg)` (`--font-size` would be `1rem`).

gavmck commented on src/app/heroes/heroes.component.css:
```diff
@@ -0,0 +1,90 @@
+        width: 15em;
+    }
+    
+    input {
```
Looks like we've got a bunch of repeating input styles that could become global

gavmck commented on src/app/heroes/heroes.component.css:
```diff
@@ -0,0 +1,90 @@
+    }
+    
+    .add-button {
+    padding: .5rem 1.5rem;
```
Tabbing

gavmck commented on src/app/heroes/heroes.component.html:
```diff
@@ -0,0 +1,19 @@
+<h2>My Heroes</h2>
+<div>
```
be nice if this submitted instead of click event

gavmck reviewed (changes requested):
Nice! Dropped a few bits in here, mostly code style stuff. I'll try and pull it down and have a play in the browser.

DanWebb commented on src/app/hero-detail/hero-detail.component.css:
```diff
@@ -0,0 +1,25 @@
+        border-radius: 4px;
+        font-size: 1rem;
+    }
+    button:hover {
```
The hover colour of these buttons in particular has an inaccessible colour contrast of 1.44

One way to see this in the chrome browser is if you bring up the browser dev tools and force the hover state on the button:

<img width="831" alt="Screenshot 2023-02-01 at 04 58 22" src="https://user-images.githubusercontent.com/5038459/215893239-60c804ed-fa49-46d0-b359-6bae7a687a8c.png">

Then click on the `color` of the button:

<img width="355" alt="Screenshot 2023-02-01 at 04 58 46" src="https://user-images.githubusercontent.com/5038459/215893485-bbab6585-e9d1-4b9a-9641-f88faf2f94bc.png">

You can play around with the colours right in the browser to see which combo of background + foreground (`color`) works.

Related reading: https://www.a11yproject.com/posts/what-is-color-contrast

DanWebb commented on src/app/hero-detail/hero-detail.component.css:
```diff
@@ -0,0 +1,25 @@
+    /* HeroDetailComponent's private CSS styles */
+    label {
```
I think everything in this file should be [globally applied styles](https://angular.io/guide/component-styles#external-and-global-style-files).

At the moment the styles for base elements like a button, label or an input is introducing more code than necessary and little inconsistencies.

It'd be good to provide a set of base/default styles that these base elements inherit then they can be overridden where necessary.

DanWebb commented on src/app/hero.service.ts:
```diff
@@ -0,0 +1,106 @@
+import { Injectable } from '@angular/core';
```
It'd be nice to see the service files a bit more organised, at the moment the root of the `/src` directory is looking like a bit of a dumping ground.

It'd be cleaner if only the app components were left in the root, then the services can be moved to another folder, maybe a structure like `/services/hero/hero.service.ts` would be good?

DanWebb reviewed (changes requested):
Hey Alijah 👋 excellent work, functionally the app works well 💪 

I've listed a few minor issues, one a11y but the rest is mainly around the organisation of code – stuff that'll come back to bite if left unattended. I think it'd could be helpful to run through these and start getting a feel for refactoring code. See what you think!

alijah-etch committed:
Changes based off of PR feedback

Made some refactoring changes and css and other changes based off of feedback received on the PR.

alijah-etch committed:
CSS to BEM and SCSS Changes

Added search-result and heroes-menu components and changed them to SCSS and BEM.

alijah-etch committed:
Fixed search-result scss/html

Followed Gav's comments to fix up the search-result scss/html.

alijah-etch committed:
testing: signing commits
